pip/pip-331.md

```diff
@@ -0,0 +1,129 @@
+
+# Motivation
+
+The server and client sides of the Pulsar function use protobuf for decoupling. In principle, the language supported by protobuf can be supported by the pulsar function, now Pulsar provided the java, python and golang function client, but there are still many languages that are not supported.
+
+Before all language adaptations are completed (and it's almost entirely certain to be impossible), users cannot write pulsar function in their familiar languages.
+
+# Goals
+
+## In Scope
+
+Other languages, as long as their code can be compiled into WASM bytecode (such as Rust/golang/C++), users can use these languages to write pulsar function.
+
+## Out of Scope
+
+All existing abilities of the Java pulsar function client are not reimplemented, the WASM Pulsar functions is under the Java Pulsar functions.
+
+Due to the strict requirements of WASM on parameter types and for simplicity reasons, types other than `java.lang.Long` are not used as parameters or return value.
+
+# High Level Design
+
+```mermaid 
+flowchart LR;
+
+    subgraph develop
+        direction TB
+        SourceCode ==> |"CompileToWASM"| WasmFile ==> |"RenameFile"| MoveToTheResourceDirectory ==> UnitTest
+    end
+
+    subgraph runtime
+        direction TB
+        PulsarFunctionJava ==> |"LoadFromResource"| TheWasmFile ==> |"Invoke"| TheSourceCode
+    end
+    
+    develop --> runtime
+```
+
+# Detailed Design
+
+## Design & Implementation Details
+
+1. add `WasmLoader` to load WASM file and provide the WASM function to java, also provide the java function to WASM if we need.
+
+2. add `AbstractWasmFunction` and `AbstractWasmWindowFunction` as the core interface of the WASM function api.
+
+```java
+public abstract class AbstractWasmFunction<X, T> extends WasmLoader implements Function<X, T> {
+
+    private static final String PROCESS_METHOD_NAME = "process";
+
+    protected static final String INITIALIZE_METHOD_NAME = "initialize";
+
+    protected static final String CLOSE_METHOD_NAME = "close";
+
+    protected static final Map<Long, Argument<?>> ARGUMENTS = new ConcurrentHashMap<>();
+
+    @Override
+    public T process(X input, Context context) {
+        return super.getWasmExtern(PROCESS_METHOD_NAME)
+                .map(process -> {
+                    Long argumentId = callWASI(input, context, process);
+                    return doProcess(input, context, argumentId);
+                })
+                .orElseThrow(() -> new PulsarWasmException(
+                        PROCESS_METHOD_NAME + " function not found in " + super.getWasmName()));
+    }
+
+    private Long callWASI(X input,
+                          Context context,
+                          Extern process) {
+        // call WASI function
+        final Long argumentId = getArgumentId(input, context);
+        ARGUMENTS.put(argumentId, new Argument<>(input, context));
```

Comment:
Can you please elaborate more on how this handover is done?
Can WASM functions have access to Java variables? How?
Can the variables be of any type? How do you map a Java class to a WASM class?

Reply:
> Can you please elaborate more on how this handover is done?
Java passes parameter IDs(`java.lang.Long`) to the other language, and then the other language obtain the raw parameters through the parameter ID.
> Can WASM functions have access to Java variables? How?
step1: export java functions to the other language

```java
public class WasmFunction extends AbstractWasmFunction<String, String> {

    private static final Map<Long, String> RESULTS = new ConcurrentHashMap<>();

    @Override
    protected Map<String, Func> initWasmCallJavaFunc(Store<Void> store) {
        Map<String, Func> funcMap = new HashMap<>();
        funcMap.put("get_args",
                WasmFunctions.wrap(store, WasmValType.I64, WasmValType.I64, WasmValType.I32, WasmValType.I32,
                        (argId, addr, len) -> {
                            String config = "hello from java " + argId;
                            System.out.println("java side->" + config);
                            ByteBuffer buf = super.getBuffer();
                            for (int i = 0; i < len && i < config.length(); i++) {
                                buf.put(addr.intValue() + i, (byte) config.charAt(i));
                            }
                            return Math.min(config.length(), len);
                        }));
        //......
        return funcMap;
    }
}
```

step2: java call WASM, the `AbstractWasmFunction#process`

step3: call to WASM, which is the following code, this is a rust(rust can be compiled into WASM) example.

```rust
#[no_mangle]
pub unsafe extern "C" fn process(arg_id: i64) {
    //......
}
```

step4: java pass args to the other language

```rust
#[link(wasm_import_module = "pulsar")]
extern "C" {
    fn get_args(arg_id: i64, addr: i64, len: i32) -> i32;
}

#[no_mangle]
pub unsafe extern "C" fn process(arg_id: i64) {
    let mut buf = [0u8; 32];
    let buf_ptr = buf.as_mut_ptr() as i64;
    // get arg from java
    let len = get_args(arg_id, buf_ptr, buf.len() as i32);
    let java_arg = std::str::from_utf8(&buf[..len as usize]).unwrap();
    eprintln!("rust side-> recv:{}", java_arg);
    // ......
}
```

> Can the variables be of any type?
No, see the supported types in `io.github.kawamuray.wasmtime.WasmValType`
> How do you map a Java class to a WASM class?
Unable to directly map, requires serialization and deserialization.

Reply:
Serialization sounds slow, as it should be done per message, no?

Reply:
Well, WASM can easily read a `byte[]` from other languages, but it cannot directly convert it into a Java object, which is far inferior to JNI, but its cross platform aspect is much worse than WASM(Imagine, the user use JNI and C to cross platform, and when users switched to Golang, they had to do it again). Any suggestions?

Reply:
Well, WASM can easily read a `byte[]` from other languages, but it cannot directly convert it into a Java object, which is far inferior to JNI, but its cross platform aspect is much worse than WASM(Imagine, the user use JNI and C to cross platform, and when users switched to Golang, they had to do it again). Any suggestions?

Reply:
Do you have any idea what is the performance implications for this step? Comparing normal java function throughput compared with this WASM function?

Reply:
@lhotari @merlimat Serializing/deserializing each message - isn't this a killer for a function runtime? Or is it legit?

Reply:
> Do you have any idea what is the performance implications for this step? Comparing normal java function throughput compared with this WASM function?

I know this will bring a significant performance drop, but besides serialization/deserialization, I can't think of any way to convert byte[] into a Java object.